core/src/internal/event_system.c

```diff
@@ -0,0 +1,333 @@
+	{
+		/* Codes_SRS_EVENTSYSTEM_26_008: [ This function shall call all registered callbacks on a seperate thread. ] */
+		THREADAPI_RESULT result = ThreadAPI_Create(&event_systen->callback_thread, callback_thread_main_func, (void*)event_systen);
+		/* Codes_SRS_EVENTSYSTEM_26_013: [ Should the worker thread ever fail to be created or any internall callbacks fail, failure will be logged and no further callbacks will be called during gateway's lifecycle. ] */
+		if (result != THREADAPI_OK)
+			event_systen->is_errored = 1;
+	}
+
+	Unlock(event_systen->internal_change_lock);
+}
+
+static void add_to_thread_queue(EVENTSYSTEM_HANDLE event_system, THREAD_QUEUE_ROW* row)
+{
+	Lock(event_system->thread_queue_lock);
+
+	list_add(event_system->thread_queue, row);
+	Condition_Post(event_system->thread_queue_condition);
+
+	Unlock(event_system->thread_queue_lock);
+}
+
+static THREAD_QUEUE_ROW* get_from_thread_queue(EVENTSYSTEM_HANDLE event_system, int timeout_ms)
+{
+	THREAD_QUEUE_ROW* row = NULL;
+	
+	Lock(event_system->thread_queue_lock);
+	
+	LIST_ITEM_HANDLE node = list_get_head_item(event_system->thread_queue);
+	if (node == NULL && event_system->delay_when_queue_empty)
+	{
+		Condition_Wait(event_system->thread_queue_condition, event_system->thread_queue_lock, timeout_ms);
+		node = list_get_head_item(event_system->thread_queue);
+	}
+
+	/* Condition might have timed out or we don't have delayed destroy so the node can be NULL */
+	if (node != NULL) {
+		row = (THREAD_QUEUE_ROW*)list_item_get_value(node);
+		list_remove(event_system->thread_queue, node);
+	}
+	
+	Unlock(event_system->thread_queue_lock);
+	
+	return row;
+}
+
+static void destroy_thread_row(THREAD_QUEUE_ROW* row)
+{
+	if (row != NULL)
+		VECTOR_destroy(row->callbacks);
+	free(row);
+}
+
+static int callback_thread_main_func(void* event_system_param)
+{
+	EVENTSYSTEM_HANDLE event_system = (EVENTSYSTEM_HANDLE)event_system_param;
+	THREAD_QUEUE_ROW* row;
+	while ((row = get_from_thread_queue(event_system, THREAD_EMPTY_QUEUE_TIMEOUT_MS)) != NULL)
+	{
+		size_t vector_size = VECTOR_size(row->callbacks);
+		/* Codes_SRS_EVENTSYSTEM_26_006: [ This function shall call all registered callbacks for the given GATEWAY_EVENT. ] */
+		/* Codes_SRS_EVENTSYSTEM_26_009: [ This function shall call all registered callbacks in First - In - First - Out order in terms registration. ] */
+		for (size_t i = 0; i < vector_size; i++)
+		{
+			GATEWAY_CALLBACK callback = *((GATEWAY_CALLBACK*)VECTOR_element(row->callbacks, i));
+			/* Codes_SRS_EVENTSYSTEM_26_010: [ The given GATEWAY_CALLBACK function shall be called with proper GATEWAY_HANDLE and GATEWAY_EVENT as function parameters coresponding to the gateway and the event that occured. ] */
+			callback(row->gateway, row->event_type, NULL);
+		}
+	}
+
+	Lock(event_system->internal_change_lock);
+
+	event_system->callback_thread = NULL;
+
+	Unlock(event_system->internal_change_lock);
+
+	ThreadAPI_Exit(THREADAPI_OK);
```

Comment:
Probably yes, but the callback function requires a return value, even if the API itself requires call to exit so it's provided.